pennylane/tensorbox/tf_box.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the TensorFlowBox implementation of the TensorBox API.
+"""
+import tensorflow as tf
+
+
+try:
+    from tensorflow.python.eager.tape import should_record_backprop
+except ImportError:  # pragma: no cover
+    from tensorflow.python.eager.tape import should_record as should_record_backprop
+
+
+import pennylane as qml
+
+
+class TensorFlowBox(qml.tensorbox.TensorBox):
+    """Implements the :class:`~.TensorBox` API for TensorFlow tensors.
+
+    For more details, please refer to the :class:`~.TensorBox` documentation.
+    """
+
+    def __len__(self):
+        if isinstance(self.data, tf.Variable):
+            return len(tf.convert_to_tensor(self.data))
+
+        return super().__len__()
+
+    @staticmethod
+    def astensor(tensor):
+        return tf.convert_to_tensor(tensor)
+
+    def cast(self, dtype):
+        return TensorFlowBox(tf.cast(self.data, dtype))
+
+    @staticmethod
+    def concatenate(values, axis=0):
+
+        if axis is None:
+            # flatten and then concatenate zero'th dimension
+            # to reproduce numpy's behaviour
+            tensors = [tf.reshape(TensorFlowBox.astensor(t), shape=[-1]) for t in TensorFlowBox.unbox_list(values)]
+            # TODO: error is raised when dtypes are not the same

```

Comment:
This is the dtype problem: `tf.concat` complains if the tensors do not have the same type. But which one is the correct one here?